### ISSUE TITLE
Comparing dictionaries compares values

### ISSUE THREAD

**Issue by bitwes** [godotengine/godot#67153]
### Godot version

4.0 Beta 2

### System information

MacOS 10.15.7

### Issue description

In 3.x dictionaries were compared by reference, but it appears they are compared by value in 4.0.  The documentation indicates they are still compared by reference.  Is this documentation wrong or has this not been implemented yet?

### Steps to reproduce

The documentation says the following will print `false` but it prints `true`.
<img width="475" alt="image" src="https://user-images.githubusercontent.com/1156520/194777043-f90a1438-dac6-4e21-9c4b-ff5e2a070e77.png">

```
extends SceneTree

func _init():
    var dict1 = {"a": 1, "b": 2, "c": 3}
    var dict2 = {"a": 1, "b": 2, "c": 3}
    print(dict1 == dict2)
    quit()
```

**Comment by KoBeWi** [comment]
I kind of clarified this in #67001 in Dictionary `==` operator. But yeah, the description is outdated.

**Comment by bitwes** [comment]
I see.  So it will use `==` operator on everything in the dictionary, so it should work recursively and with arrays as well?

Is there a way to compare them by reference?  I tried `weakref` and `get_instance_id` which don't work/exist on dictionaries.  I didn't see any methods that would work either.  Since they are pass-by-reference, there should probably be a way to compare them by reference.  It might be more intuitive (at least for 3.x users) if `==` compared the reference and a `compare` method was added.  Comparing large dictionaries by value could be expensive especially if you are just trying to see if you already have a reference to it.

This might also apply to Arrays since they changed from pass-by-value to pass-by-ref.  I could see needing to check if you had the same reference to an array instead of checking the values.  Especially on large arrays.

**Comment by KoBeWi** [comment]
Dictionaries have `id()` method that can be used for comparing references. We could maybe expose it and mention in the docs. I think it was brought up when the comparing behavior was changed, but I don't remember why it wasn't done 🤔

**Comment by Mickeon** [comment]
The more I look into it, the more I feel like the equality by value is... a bit unwarranted. As much sense as it may make at first glance, it starts getting messy as one gets the feel for GDScript and other concepts (passed by reference VS. passed by value). For other languages, the problem is a bit more exacerbated, because they do not behave that way.  Not to mention that, for this to make complete sense, any Object should be also compared by value. I feel like having a method for this comparison by value would be somewhat nicer.

But, okay, it's fine, it's alright.
I am however running into issues exposing `id()` to GDScript. It returns a pointer, internally, so... Hmm...

**Comment by bitwes** [comment]
It looks like `is_same` is how to compare dictionaries and arrays by reference in 4.0.  I think this has been fixed by #71758.  

https://docs.godotengine.org/en/latest/classes/class_@globalscope.html#class-globalscope-method-is-same
